compiler/compiler.go

```diff
@@ -277,19 +255,98 @@ func WritePkgCode(pkg *Archive, dceSelection map[*Decl]struct{}, gls goLinknameS
 	return nil
 }
 
+type serializableArchive struct {
+	ImportPath   string
+	Name         string
+	Imports      []string
+	ExportData   []byte
+	Declarations []*Decl
+	IncJSCode    []byte
+	FileSet      []byte
+	Minified     bool
+	GoLinknames  []GoLinkname
+	BuildTime    time.Time
+}
+
 // ReadArchive reads serialized compiled archive of the importPath package.
-func ReadArchive(path string, r io.Reader) (*Archive, error) {
+//
+// The given srcModTime is used to determine if the archive is out-of-date.
+// If the archive is out-of-date, the returned archive is nil.
+// If there was not an error, the returned time is when the archive was built.
+//
+// The imports map is used to resolve package dependencies and may modify the
+// map to include the package from the read archive. See [gcexportdata.Read].
+func ReadArchive(importPath string, r io.Reader, srcModTime time.Time, imports map[string]*types.Package) (*Archive, time.Time, error) {
+	var sa serializableArchive
+	if err := gob.NewDecoder(r).Decode(&sa); err != nil {
+		return nil, time.Time{}, err
+	}
+
+	if srcModTime.After(sa.BuildTime) {
+		// Archive is out-of-date.
+		return nil, sa.BuildTime, nil
+	}
+
 	var a Archive
-	if err := gob.NewDecoder(r).Decode(&a); err != nil {
-		return nil, err
+	fset := token.NewFileSet()
+	if len(sa.ExportData) > 0 {
+		pkg, err := gcexportdata.Read(bytes.NewReader(sa.ExportData), fset, imports, importPath)
+		if err != nil {
+			return nil, sa.BuildTime, err
+		}
+		a.Package = pkg
+	}
+
+	if len(sa.FileSet) > 0 {
+		a.FileSet = token.NewFileSet()
+		if err := a.FileSet.Read(json.NewDecoder(bytes.NewReader(sa.FileSet)).Decode); err != nil {
+			return nil, sa.BuildTime, err
+		}
 	}
 
-	return &a, nil
+	a.ImportPath = sa.ImportPath
+	a.Name = sa.Name
+	a.Imports = sa.Imports
+	a.Declarations = sa.Declarations
+	a.IncJSCode = sa.IncJSCode
+	a.Minified = sa.Minified
+	a.GoLinknames = sa.GoLinknames
+	return &a, sa.BuildTime, nil
```

Comment:
If we think some sort of timestamp is relevant to the consumers of the archive, we should make it a part of the struct, instead of returning it separately. See also the comment below that it should be a derivative of SrcModTime.

Reply:
My thought was, if we return a nil archive and nil error, there is no explanation for why it returned a nil archive, but if we needed to we could see the buildTime and determine that the archive was just out-of-date. Right now, that's just used for some logging right after load and then isn't used anymore.

I was trying to separate what is needed for reading and writing an Archive and what the Archive used for once loaded. I guess I figured that the Archive didn't need to know it's build time and that's just part of ensuring we don't load an out-of-date archive. Kinda like the build time is a little more resilient file mod time that can't be changed by a move or touch.

Seems like I might be thinking about it wrong. I don't mind undoing and putting the BuildTime field back on Archive. Since you definitely understand this better than I do, I'll do what you think makes the most sense.

Reply:
Right, my comment was in conjunction with the one above, and your point about side effects in the imports map is a valid one. Ideally, each archive would decode gcexportdata into its own imports map, which could be merged downstream, but the interface of `gcexportdata.Read` doesn't really make it easy.

Given the constraints, the current approach makes sense.